File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l, r = 0, len(nums) - 1
        
        while (l <= r):
            m = l + (r - l) / 2
            
            if nums[m] == target:
                return m
            
            if nums[m] >= nums[l]:
                if target >= nums[l] and target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                if target > nums[m] and target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
        
        return -1
        

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int l = 0, r = nums.length - 1;

        while (l <= r) {
            int m = l + ((r - l) / 2);

            if (nums[m] == target) {
                return m;
            }
            
            // left sorted portion
            if (nums[l] <= nums[m]) {
                if (target < nums[l]) {
                    l = m + 1;
                } else if (target > nums[m]) {
                    l = m + 1;
                } else {
                    r = m - 1;
                }
            } else { // right sorted portion
                if (target > nums[r]) {
                    r = m - 1;
                } else if (target < nums[m]) {
                    r = m - 1;
                } else {
                    l = m + 1;
                }
            }
        }

        return -1;
    }
}